### DISCUSSION TITLE
Data Scraper

### DISCUSSION THREAD

**Discussion by 5h3rr1ll** [KitaPlatzZentrale/kpz#109] [Ideas]
## Problem
We want to make sure that we always have updated data.

## Solution 1
Data Scraper that runs nightly

**Comment by 5h3rr1ll** [comment]
So, with just one request, we can get all the daycare centers in Berlin from Berlin.de. The one request could be made once every night to get the data first. We could then see where we store it; it could simply be a text file stored in an S3 bucket. Then it's about checking our data status. Daily changes are less likely to be expected, but it would still be nice if we had a smart logic that compares the nightly fed data with our database. However, we only need to update the entries that have changed. In my opinion, simply an "updated_at" timestamp is enough to mark the update so that we can use the version parameter for different versions of our JSON. If we optimize the logic, we won't have any problems with time-outs. Our DB won't be so cluttered, as we will only ever have as many entries as there are daycare centers in Berlin. This should also keep performance optimal, as we never have to iterate through more entries than necessary. However, we store 1200 new childcare centers every day. In that case, we will iterate through almost half a million entries after a year, even though only a few true/false values have changed, and maybe a few pieces of information whose historical value has no value for us at all. ☺️ Do you know what I mean? ☺️

**Comment by Darjusch** [comment]
First of all, it's nice that we can discuss it here! 
I suggest that we use headers for each topic of discussion so we have it sorted otherwise it's a big wall of text.

Short recap.
### Current form
We fetch kitas from Berlin.de
We check if data is new by creating a hash and comparing it
For each kita, we fetch the detailed view by UUID
We save it to the database

### Suggestion
We fetch kitas from Berlin.de
We check for each single kita if the data is new - for this, we need to fetch all kitas from the db and compare all availabilities
Then we save them to the database 

### Biggest change
Most time we save here by not fetching the detailed view and saving the "extra" data. This data includes opening hours and special information about the kita.
Instead of adding data we should also only update current entries or add new entries if they don't exist yet. 


### Repurpose Version
We should replace the purpose of version with created_at and use version only for changes in the data structure of the json.

Do you agree with this?

**Comment by Darjusch** [comment]
Current Solution: 
Batching the requests into 100's using Promise.all and map.
Overwriting the old Documents with the updated ones. 
Not using Version anymore but createdAt.

Performance is at around 10 Seconds local
